end_to_end/end_to_end_suite_test.go

```diff
@@ -2657,4 +2657,231 @@ LANGUAGE plpgsql NO SQL;`)
+				//In 6X we also should backup siblings
+				Expect(metadataFileContents).To(ContainSubstring("ALTER TABLE public.t_part ALTER PARTITION FOR (RANK(1)) ADD PARTITION part_d VALUES ('d'::text);"))
+				//We shouldn't create explicit create leaf partiton if it's parent is not in extension
+				Expect(metadataFileContents).ToNot(ContainSubstring("a_part"))
+				Expect(metadataFileContents).ToNot(ContainSubstring("b_part"))
+			}
+
+			testhelper.AssertQueryRuns(restoreConn, `CREATE EXTENSION test_ext6;`)
+			defer testhelper.AssertQueryRuns(restoreConn, `DROP EXTENSION test_ext6 CASCADE;`)
+
+			gprestore(gprestorePath, restoreHelperPath, timestamp, "--backup-dir", backupDir, "--redirect-db", "restoredb")
+
+			if backupConn.Version.AtLeast("7") {
+				assertDataRestored(restoreConn, map[string]int{
+					"public.part_a": 10,
+					"public.part_c": 30,
+				})
+			} else {
+				assertDataRestored(restoreConn, map[string]int{
+					"public.t_part_1_prt_test_part_2_prt_a_part": 10,
+					"public.t_part_1_prt_test_part_2_prt_b_part": 20,
+					"public.t_part_1_prt_1_2_prt_part_c":         30,
+					"public.t_part_1_prt_1_2_prt_part_d":         40,
+				})
+			}
+
+			assertArtifactsCleaned(restoreConn, timestamp)
+		})
+
+		It("normally handles the '--exclude-table' during backup", func() {
+			if backupConn.Version.AtLeast("7") {
+				testhelper.AssertQueryRuns(backupConn, `
+					CREATE TABLE t_part_1_prt_1_2_prt_part_c PARTITION OF t_part_1_prt_1 FOR VALUES IN ('c');
+					CREATE TABLE t_part_1_prt_1_2_prt_part_d PARTITION OF t_part_1_prt_1 FOR VALUES IN ('d');`)
+			} else {
+				testhelper.AssertQueryRuns(backupConn, `
+					ALTER TABLE t_part ALTER PARTITION FOR (RANK(1)) ADD PARTITION part_c VALUES ('c');
+					ALTER TABLE t_part ALTER PARTITION FOR (RANK(1)) ADD PARTITION part_d VALUES ('d');`)
+			}
+
+			timestamp := gpbackup(gpbackupPath, backupHelperPath, "--backup-dir", backupDir, "--leaf-partition-data",
+				"--exclude-table", "public.t_part_1_prt_1_2_prt_part_c")
+
+			metadataFileContents := string(getMetdataFileContents(backupDir, timestamp, "metadata.sql"))
+
+			if backupConn.Version.AtLeast("7") {
+				Expect(metadataFileContents).To(ContainSubstring("ALTER TABLE ONLY public.t_part_1_prt_1 ATTACH PARTITION public.t_part_1_prt_1_2_prt_part_d FOR VALUES IN ('d');"))
+				Expect(metadataFileContents).ToNot(ContainSubstring("ALTER TABLE ONLY public.t_part_1_prt_1 ATTACH PARTITION public.t_part_1_prt_1_2_prt_part_c FOR VALUES IN ('c');"))
+			} else {
+				Expect(metadataFileContents).To(ContainSubstring("ALTER TABLE public.t_part ALTER PARTITION FOR (RANK(1)) ADD PARTITION part_d VALUES ('d'::text);"))
+				Expect(metadataFileContents).ToNot(ContainSubstring("ALTER TABLE public.t_part ALTER PARTITION FOR (RANK(1)) ADD PARTITION part_c VALUES ('c'::text);"))
+			}
+
+			assertArtifactsCleaned(restoreConn, timestamp)
+		})
+
+		It("backup the default partition the last one", func() {
```

Comment:
We broke a lot of spears around how gpbackup should back up extension with default partition, but we have only one test for now. I think we should test the desirable behavior when default partition is in extension or added later and do it for both, 6x and 7x.

Reply:
As I understand it, we can only handle the case when the default partition was added by the user after the extension was created. This is what is being tested in this test. The behavior for 6X and 7X is actually not very different if there is data in the default partition. I will add a test for 7X.

Reply:
In 7X, we need to split the default partition only if the default partition contains data. Since we restore data only after creating a schema in 7X, we do not care about the order of restoration of partitions. Therefore, I think that the test for 7X does not make sense.

Reply:
One of the goals of tests is to nail the desired behavior. We have multiple interesting cases. Some of them may be intuitive, but it doesn't mean that we shouldn't test them. Default partition may be inside extension(1) or added later (2). Additional partition may be added to extension with(3) or without(4) default partition. This may be done differently for 6x(5) and 7x(6). Most of code which handles this logic is added by this patch, but anyway, we should nail the desired behavior.